language/move-binary-format/src/file_format.rs

```diff
@@ -1668,9 +1668,8 @@ impl Bytecode {
     /// Return the successor offsets of this bytecode instruction.
     pub fn get_successors(pc: CodeOffset, code: &[Bytecode]) -> Vec<CodeOffset> {
         assert!(
-            // The program counter could be added to at most twice and must remain
-            // within the bounds of the code.
-            pc <= u16::max_value() - 2 && (pc as usize) < code.len(),
+            // The program counter must remain within the bounds of the code
+            pc < u16::MAX && (pc as usize) < code.len(),
```

Comment:
Won't we get an overflow at 1682 (`let next_pc = pc + 1`) or 1689 `v.push(pc + 1)` without the `- 2`?

I see that you are trying to test this indirectly by invoking the verifier below, but I think it would be better to directly test `get_succesors` with (e.g.) `pc = u16::MAX - 1`.

Reply:
Got it, thanks for the detailed explanation!

Definitely important to fix the panic, but I am wondering if we can do it in a more future-proof way. I buy the argument that the usage of `get_successors` in `VMControlGraph::new` is safe, but because `get_successors` is a public function, it's easy to imagine it being called in a different context where the same assumptions don't hold (e.g., it also gets called from `tools/move-coverage/src/summary.rs`, could be called in a future verifier pass). Couple of ideas on this, but no strong prefs:

- Move `get_successors` inside of `VMControlFlowGraph` and make it private
- Move `get_successors` inside of `VMControlFlowGraph` and check `is_end_of_block` as a precondition
- Have `get_successors` return an empty vec if `pc > u16::max_value() - 2`